### PR TITLE
added python bindings

### PR DESCRIPTION
**Pull request by caoliheng**

### CODE REVIEW

File: include/imu-core/imu_3DM_GX3_25.hpp [hunk 1]
```diff
@@ -315,4 +315,4 @@ class Imu3DM_GX3_25: public ImuInterface
 
 } // namespace imu_3DM_GX3_25
 } // namespace imu_core
-#endif // IMU_3DM_GX3_25_HPP
+#endif // IMU_3DM_GX3_25_HPP
```
Comment (jviereck) [inline include/imu-core/imu_3DM_GX3_25.hpp:318]: Why did you remove the newline here?

File: include/imu-core/imu_interface.hpp [hunk 2]
```diff
@@ -7,6 +7,8 @@
 #include <Eigen/Eigen>
 #include "real_time_tools/usb_stream.hpp"
 
+
+
```
Comment (jviereck) [inline include/imu-core/imu_interface.hpp:11]: If possible, try not to commit whitespace/newline changes until they are necessary / you are doing a code formation.

File: srcpy/imu_core_cpp.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+#include <pybind11/eigen.h>
+
+// #include <imu-core/imu_3DM_GX3_25.hpp>
+#include <imu-core/imu_interface.hpp>
+#include <imu-core/imu_3DM_GX3_25.hpp>
+
+namespace py = pybind11;
+
+// lets "i"_a be the same as py::arg("i") for keyword arguments
+using namespace pybind11::literals;
```
Comment (jviereck) [inline srcpy/imu_core_cpp.cpp:16]: Didn't new about this one - nice :)

File: srcpy/imu_core_cpp.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+class PyImuInterface : public ImuInterface{
+public:
+  using ImuInterface::ImuInterface;
+
+  bool initialize() override { 
+    // deprecated in newer versions of pybind11, just rename PYBIND11_OVERLOAD_PURE --> PYBIND11_OVERRIDE_PURE without changing arguments
+    PYBIND11_OVERLOAD_PURE( 
+      bool, // return type
+      ImuInterface, // parent class
+      initialize, // name of function
```
Comment (jviereck) [inline srcpy/imu_core_cpp.cpp:31]: "name of function" -> "name of constructor"?

File: srcpy/imu_core_cpp.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    .def("start_streaming_data", &Imu3DM_GX3_25::start_streaming_data)
+    .def("stop_streaming_data", &Imu3DM_GX3_25::stop_streaming_data)
+    .def_static("reading_loop_helper", &Imu3DM_GX3_25::reading_loop_helper, "object"_a)
+    .def("reading_loop", &Imu3DM_GX3_25::reading_loop)
+    .def("stop_reading_loop", &Imu3DM_GX3_25::stop_reading_loop)
+    .def("receive_acc_gyro", &Imu3DM_GX3_25::receive_acc_gyro, "stream_data")
+    .def("receive_stab_acc_gyro_magn", &Imu3DM_GX3_25::receive_stab_acc_gyro_magn)
+    .def("receive_acc_gyro_rot_mat", &Imu3DM_GX3_25::receive_acc_gyro_rot_mat)
+    .def("receive_quaternion", &Imu3DM_GX3_25::receive_quaternion);
+}
```
Comment (jviereck) [inline srcpy/imu_core_cpp.cpp:78]: Missing newline.

File: srcpy/imu_core_cpp.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    .def("capture_gyro_bias", &Imu3DM_GX3_25::capture_gyro_bias)
+    .def("start_streaming_data", &Imu3DM_GX3_25::start_streaming_data)
+    .def("stop_streaming_data", &Imu3DM_GX3_25::stop_streaming_data)
+    .def_static("reading_loop_helper", &Imu3DM_GX3_25::reading_loop_helper, "object"_a)
+    .def("reading_loop", &Imu3DM_GX3_25::reading_loop)
+    .def("stop_reading_loop", &Imu3DM_GX3_25::stop_reading_loop)
+    .def("receive_acc_gyro", &Imu3DM_GX3_25::receive_acc_gyro, "stream_data")
+    .def("receive_stab_acc_gyro_magn", &Imu3DM_GX3_25::receive_stab_acc_gyro_magn)
+    .def("receive_acc_gyro_rot_mat", &Imu3DM_GX3_25::receive_acc_gyro_rot_mat)
+    .def("receive_quaternion", &Imu3DM_GX3_25::receive_quaternion);
```
Comment (jviereck) [inline srcpy/imu_core_cpp.cpp:77]: I was not expecting to see all these methods here but I guess it doesn't hurt to have them exposed.

**Review by jviereck — COMMENTED** [review]
Looks good overall. Will do another review once the python code is properly running.

File: include/imu-core/imu_3DM_GX3_25.hpp (hunk 1 above)
Comment (caoliheng) [reply include/imu-core/imu_3DM_GX3_25.hpp:318]: accident

File: pydemo/test.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import signal
+import sys
+import time
+import numpy as np
+import imu_core.imu_core_cpp as IMU
+
+
+def handle_signal(s, f):
+    # this way is better then the commented version because you want
+    # to force it to call destructor immediately
```
Comment (jviereck) [inline pydemo/test.py:10]:
```
    # this way is better then the commented version because you want
    # to force it to call destructor immediately
```

Should this comment be removed?

File: pydemo/test.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    # For example, in ipython, you would have to exit the ipython
+    # instead completely instead of just ctrl-c to exit the script
+    print("\nExiting!\n")
+    global imu
```
Comment (jviereck) [inline pydemo/test.py:14]: Please put the `global` defs at the beginning of the function to make it easier to spot.

File: pydemo/test.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    del imu
```
Comment (jviereck) [inline pydemo/test.py:15]: You really have to all `del` on the IMU? There is no way to run the destructor once the IMU object gets garbage collected? Did you check the pybind docs?

File: pydemo/test.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+else:
+    # wrong number of arguments
+    sys.exit(f"Wrong number of arguments: expected 1, received {len(argv)}")
+
+
+imu = IMU.Imu3DM_GX3_25(port, True)
+
+imu.initialize()
+
+np.set_printoptions(precision=3, suppress=True, sign=' ', floatmode='fixed')
```
Comment (jviereck) [inline pydemo/test.py:39]: Put the `np.set_printoptions` at the beginning after the import please as you will modify it globally.

File: src/imu_3DM_GX3_25.cpp [hunk 11]
```diff
@@ -75,12 +75,26 @@ bool Imu3DM_GX3_25::open_usb_port(int bauderate)
...
-  if(!usb_stream_.read_device(msg.reply_, stream_mode))
+  ssize_t read_bytes = usb_stream_.read_maybe_device(msg.reply_, stream_mode);
+  if(read_bytes < 0)
   {
     rt_printf("Imu3DM_GX3_25::receive_message(): [Error] "
               "message badly received\n");
   }
 
+  // rt_printf("\nBREAK debug received message of size %ld, expected %ld\n\n", asdf, msg.reply_.size());
+  if(read_bytes != static_cast<ssize_t>(msg.reply_.size()))
```
Comment (jviereck) [inline src/imu_3DM_GX3_25.cpp:86]: would it make sense to change this into a `while` loop? `while (total_bytes < expected_bytes):`

File: src/imu_3DM_GX3_25.cpp [hunk 12]
```diff
@@ -164,7 +178,7 @@ bool Imu3DM_GX3_25::read_misaligned_msg_from_device(ImuMsg& msg)
                 "Could not read fragment.\n");
       return false;
     }
-    // we shift the end of teh message to the beginning
+    // we shift the end of the message to the beginning
```
Comment (jviereck) [inline src/imu_3DM_GX3_25.cpp:181]: Thanks for doing all these small typo fixes!

**Review by jviereck — CHANGES_REQUESTED** [review]
Small comments, but overall this looks good!

File: pydemo/test.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+if __name__ == "__main__":
+    def handle_signal(s, f):
+        global keep_running
+        keep_running = False
+        
+    signal.signal(signal.SIGINT, handle_signal)
+
+    if len(sys.argv) == 1:
+        # default port
+        # port = "/dev/ttyACM0"
```
Comment (jviereck) [inline pydemo/test.py:18]: nit, can you remove the commented out line here?

File: pydemo/test.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    print("\nExiting!")    
+    # C++ destructor will get automatically called, but not necessarily when the 
+    # script ends.
+    # This can lead to problems when using ipython if you don't exit out of ipython
+    # fully before calling this script again.
+    # Eg., if you run ipython, run test.py, ctrl-c, the destructor won't run. 
+    # This is because ipython keeps variables.
+    # del imu forces the destructor to run.
+    del imu
+    sys.exit(0)
```
Comment (jviereck) [inline pydemo/test.py:47]: nit, missing newline

File: pydemo/test.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        time.sleep(0.05)
+
+    print("\nExiting!")    
+    # C++ destructor will get automatically called, but not necessarily when the 
+    # script ends.
+    # This can lead to problems when using ipython if you don't exit out of ipython
+    # fully before calling this script again.
+    # Eg., if you run ipython, run test.py, ctrl-c, the destructor won't run. 
+    # This is because ipython keeps variables.
+    # del imu forces the destructor to run.
```
Comment (jviereck) [inline pydemo/test.py:45]: This comment is great!

**Review by jviereck — APPROVED** [review]
Small nits to address, otherwise this is ready for landing from my side.

**Comment by jviereck** [comment]
Thanks a lot for working on this @caoliheng !
